---
title: "Single-cell RNA-seq analysis: Quality Control"
author: "Csilla Varnai"
date: "2022-09-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
```


You will be analysing a scRNA-seq dataset of peripheral blood cells,
generated by the 10X Genomics technology, using the Seurat R package().

This activity follows the PBMC tutorial using Seurat, as can be found here:
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html




# 1. Input files

First, let us input the scRNA-seq data, using the Read10X() function.
This will input all the three input files (count matrix, gene IDs and cell barcodes)
and stores it in a Seurat object.

```{r}
pbmc.data <- Read10X(data.dir = "./filtered_gene_bc_matrices/hg19/")
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
```

Let us inspect the data.  First, the Seurat object:
```{r}
pbmc
```

The count matrix is stored as pbmc.data.

Any metadata associated with the SeuratObject can be viewed the following way:
```{r}
head(pbmc@meta.data, 5)
```

The [[ ]] operator can be used to add further metadata, such as patient ID, batch or replicate ID, any information about the biological samples.


# 2. Quality control and filtering


For each cell, let us calculate the percentage of mitochondrial reads in the data using the PercentageFeatureSet() function, and store it as percent.mt using the [[ ]] operator.

```{r}
# The [[ operator can add columns to object metadata.
pbmc[["percent.mt"]] <- PercentageFeatureSet(pbmc, pattern = "^MT-")

# Show QC metrics for the first 5 cells
head(pbmc@meta.data, 5)
```

Use the FeatureScatter() function to plot the QC metrics of cells against one another.
What are these metrics and how do they correlate with one another?

```{r}
plot1 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot2 <- FeatureScatter(pbmc, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot3 <- FeatureScatter(pbmc, feature1 = "nFeature_RNA", feature2 = "percent.mt")

plot1 + plot2 + plot3
```


Next, let's plot the distributions of QC metrics as a violin plot, using the VlnPlot() function.

```{r}
VlnPlot(pbmc, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```

Apply the following filters to the data, to only keep good quality cells.

```{r}
pbmc_filtered <- subset(pbmc, subset = nFeature_RNA > 200 & nFeature_RNA < 2500 & percent.mt < 5)
```


```{r}
pbmc_filtered
```


# 3. Save the filtered data.

Finally, save the filtered SeuratObject for further analysis.

```{r}
saveRDS(pbmc_filtered, file = "pbmc3k_filtered.rds")
```

