---
title: "Single-cell RNA-seq analysis: Interpretation."
author: "Csilla Varnai"
date: "2022-09-23"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(dplyr)
library(Seurat)
library(patchwork)
library(SingleR)
library(celldex)
```


You will continue to be analysing a scRNA-seq dataset of peripheral blood cells,
generated by the 10X Genomics technology, using the Seurat R package().

This activity follows the PBMC tutorial using Seurat, as can be found here:
https://satijalab.org/seurat/articles/pbmc3k_tutorial.html

First, let us input the filtered, normalised and clustered scRNA-seq data we have previously created (relying on the first 15 principal components), using the readRDS() function.


```{r}
pbmc_clustered <- readRDS("pbmc3k_clustered_PCA_15.rds")
pbmc_clustered
```


# 1. Cluster biomarkers

We can identify cluster markers, i.e. genes that are specifically up- or down-regulated in different cell clusters, using differential expression analysis.
Take a look at the help of the FindMarkers() and FindAllMarkers() functions, and see what the options mean.
By default, FindMarkers() and FindAllMarkers() compares a cell cluster to all other cells combined.

We will run the FindAllMarkers() function to identify the cluster markers.  What do these arguments mean?

```{r}
pbmc.markers <- FindAllMarkers(pbmc_clustered, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
head(pbmc.markers)
```

Let us write the cluster markers of each cluster into a text file.

```{r}
write.table(x = pbmc.markers,
    file = "pbmc_markers_pos.tsv", append = FALSE, quote = FALSE, sep = "\t", row.names = FALSE, col.names = TRUE)
```


Let us visualise the distribution of expression levels of the top 10 cluster 0 markers in the different clusters using a violin plot, by running the VlnPlot() function.

```{r}
cluster0_top10_markers <- pbmc.markers %>%
    filter(cluster == 0) %>%
    slice_max(n = 10, order_by = avg_log2FC)
VlnPlot(pbmc_clustered, features = cluster0_top10_markers$gene)
```

Now let us plot the expression levels of cluster 0's cluster markers across all cells, in the UMAP dimensionality reduced data.

```{r}
top1_markers = pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 1, order_by = avg_log2FC)
FeaturePlot(pbmc_clustered, features = top1_markers$gene)
```

We can also plot a heatmap of expression levels of the top 10 cluster markers of all clusters, in cells ordered by cluster ID,
using the DoHeatmap() function.
```{r}
top10_markers = pbmc.markers %>%
    group_by(cluster) %>%
    slice_max(n = 10, order_by = avg_log2FC)
DoHeatmap(pbmc_clustered, features = top10_markers$gene) + NoLegend()
```


Feel free to explore other plotting features of Seurat, such as RidgePlot(), CellScatter(), and DotPlot().


# 2. Manual celltype identification

For this dataset, we can use canonical markers to annotate the clusters.

```{r}
cell_types = c("Naive CD4 T", "Memory CD4 T", "CD8 T", "B", "DC", "NK", "CD14+ Mono", "FCGR3A+ Mono", "Platelet")

canonical_markers = c("IL7R", "CCR7", "IL7R", "S100A4", "CD8A", "MS4A1", "FCER1A", "CST3", "GNLY", "NKG7", "CD14", "LYZ", "FCGR3A", "MS4A7", "PPBP")
```


```{r}
DoHeatmap(pbmc_clustered, features = canonical_markers) + NoLegend()
```

You can also plot violinplots and scatter plots with expression to help you match the expression profiles of the canonical markers to the cluster IDs.

Based on the plot, order the cell types to match the clsuters.  You will need to change the order to match the feature plots!
```{r}
# these are in the wrong order!
#new.cluster.ids <- c("Naive CD4 T", "Memory CD4 T", "CD8 T", "B", "DC", "NK", "CD14+ Mono", "FCGR3A+ Mono", "Platelet")
```

Let us set new cluster ID based on the expression of these features.

```{r}
names(new.cluster.ids) <- levels(pbmc_clustered)
pbmc_clustered_manual <- RenameIdents(pbmc_clustered, new.cluster.ids)

DimPlot(pbmc_clustered_manual, reduction = "umap", label = TRUE, pt.size = 0.5) #+ NoLegend()
```


# 3. Automated celltype identification

We will use the singleR package to automatically identify cell types by comparing the expression profiles of clusters to a reference dataset.
We will use the HumanPrimaryCellAtlasData reference dataset, which contains primarily microarray data for blood cell types, appropriate for our data.


```{r}
hpca_reference <- HumanPrimaryCellAtlasData()
head(hpca_reference)
pred <- SingleR(GetAssayData(pbmc_clustered, assay = "RNA", slot = "data"), clusters = Idents(pbmc_clustered),ref = hpca_reference, labels = hpca_reference$label.main)
```

The new cluster labels:

```{r}
pred$first.labels
```

Let us set these to be the new cluster IDs, and plot the data.

```{r}
new.cluster.ids <- pred$first.labels
names(new.cluster.ids) <- levels(pbmc_clustered)
pbmc_clustered_automated <- RenameIdents(pbmc_clustered, new.cluster.ids)
DimPlot(pbmc_clustered_automated, reduction = "umap", label = TRUE, pt.size = 0.5) #+ NoLegend()
```


